### PR TITLE
executor: remove some useless code and avoid some redundancy check

### PR DESCRIPTION
**Pull request by jackysp**
### What problem does this PR solve? 
Remove some useless code and avoid some redundancy check.

### What is changed and how it works?
Remove `finished` in all DMLs. Avoid once null check when filling the rows.

### Check List 

Tests 

 - Unit test

Code changes

 - Has exported function/method change

PTAL @coocood 

Merge it with https://github.com/pingcap/tidb-test/pull/623

### CODE REVIEW

**Comment by jackysp** [comment]
PTAL @coocood

File: table/column_test.go [hunk 1]
```diff
@@ -338,7 +338,7 @@ func (t *testTableSuite) TestGetDefaultValue(c *C) {
 				},
 			},
 			true,
-			types.Datum{},
+			types.NewIntDatum(0),
```
Comment (tiancaiamao) [inline table/column_test.go]: Is this a bug fix?

File: table/column.go [hunk 2]
```diff
@@ -339,7 +353,7 @@ func getColDefaultValue(ctx sessionctx.Context, col *model.ColumnInfo, defaultVa
 
 func getColDefaultValueFromNil(ctx sessionctx.Context, col *model.ColumnInfo) (types.Datum, error) {
 	if !mysql.HasNotNullFlag(col.Flag) {
-		return types.Datum{}, nil
```
Comment (tiancaiamao) [inline table/column.go]: `types.Datum{}` is better than `types.NewDatum(nil)`

File: table/column.go [hunk 3]
```diff
@@ -348,7 +362,10 @@ func getColDefaultValueFromNil(ctx sessionctx.Context, col *model.ColumnInfo) (t
 	}
 	if mysql.HasAutoIncrementFlag(col.Flag) {
 		// Auto increment column doesn't has default value and we should not return error.
-		return types.Datum{}, nil
+		return GetZeroValue(col), nil
+	}
+	if col.IsGenerated() {
+		return types.NewDatum(nil), nil
```
Comment (tiancaiamao) [inline table/column.go]: What if the generated column have a not null flag ?

File: executor/insert_common.go [hunk 4]
```diff
@@ -368,102 +352,115 @@ func (e *InsertValues) getColDefaultValue(idx int, col *table.Column) (d types.D
...
+		return d, nil
+	}
+	return datum, nil
+}
+
+// fillRow fills generated columns, auto_increment column and empty column.
 // For NOT NULL column, it will return error or use zero value based on sql_mode.
-func (e *InsertValues) initDefaultValues(row []types.Datum, hasValue []bool) error {
+func (e *InsertValues) fillRow(cols []*table.Column, row []types.Datum, hasValue []bool, valLen int) ([]types.Datum, error) {
 	for i, c := range e.Table.Cols() {
```
Comment (tiancaiamao) [inline executor/insert_common.go]: There are `cols` from caller and `e.Table.Cols()`, which one should we use?

File: executor/insert_common.go [hunk 5]
```diff
@@ -368,102 +352,115 @@ func (e *InsertValues) getColDefaultValue(idx int, col *table.Column) (d types.D
...
+		var err error
+		// Get the default value for all no value columns, the auto increment column is different from the others.
+		row[i], err = e.fillColValue(row[i], i, c, hasValue[i])
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+
+		// Evaluate the generated columns.
+		if c.IsGenerated() {
+			for i, expr := range e.GenExprs {
```
Comment (tiancaiamao) [inline executor/insert_common.go]: s/i/j

File: executor/insert_common.go [hunk 6]
```diff
@@ -368,102 +352,115 @@ func (e *InsertValues) getColDefaultValue(idx int, col *table.Column) (d types.D
...
+// fillColValue fills the column value if it is not set in the insert statement.
+func (e *InsertValues) fillColValue(datum types.Datum, idx int, column *table.Column, hasValue bool) (types.Datum,
+	error) {
+	if mysql.HasAutoIncrementFlag(column.Flag) {
+		d, err := e.adjustAutoIncrementDatum(datum, hasValue, column)
+		if err != nil {
+			return types.Datum{}, errors.Trace(err)
+		}
+		return d, nil
+	} else if !hasValue {
```
Comment (tiancaiamao) [inline executor/insert_common.go]: s/else if/ if

Comment (tiancaiamao) [reply executor/insert_common.go]: The "not auto increment" && "not has value case".

Comment (jackysp) [reply executor/insert_common.go]: If not auto increment and not has value, we could return the datum directly.

File: executor/insert_common.go (hunk 4 above)
Comment (jackysp) [reply executor/insert_common.go]: Good catch! I think it may contain some bugs.

File: table/column.go (hunk 3 above)
Comment (jackysp) [reply table/column.go]: It will be checked in `HandleBadNull`.

File: table/column_test.go (hunk 1 above)
Comment (jackysp) [reply table/column_test.go]: No, it a unit test for function `GetColDefaultValue`. It is changed due to the behavior of `GetColDefaultValue` was changed.

**Comment by zz-jason** [comment]
LGTM
@jackysp Please resolve the conflicts.

File: executor/insert_common.go (hunk 6 above)
Comment (tiancaiamao) [reply executor/insert_common.go]: else if can be changed to if, rest LGTM

**Comment by jackysp** [comment]
PTAL @coocood

**Comment by tiancaiamao** [comment]
LGTM

**Comment by jackysp** [comment]
PTAL @coocood

**Comment by jackysp** [comment]
PTAL @coocood

**Comment by jackysp** [comment]
PTAL @coocood

**Comment by jackysp** [comment]
PTAL @lysu

File: executor/insert_common.go [hunk 7]
```diff
@@ -368,102 +352,116 @@ func (e *InsertValues) getColDefaultValue(idx int, col *table.Column) (d types.D
...
+	error) {
+	if mysql.HasAutoIncrementFlag(column.Flag) {
+		d, err := e.adjustAutoIncrementDatum(datum, hasValue, column)
+		if err != nil {
+			return types.Datum{}, errors.Trace(err)
+		}
+		return d, nil
+	}
+	if !hasValue {
+		d, err := e.getColDefaultValue(idx, column)
```
Comment (lysu) [inline executor/insert_common.go]: if col is `generate` + `!hasValue`, this call can be eliminated?

File: executor/write_test.go [hunk 8]
```diff
@@ -635,11 +635,16 @@ commit;`
 	tk.MustQuery(`SELECT * FROM t1;`).Check(testkit.Rows("1"))
 
 	testSQL = `DROP TABLE IF EXISTS t1;
-	CREATE TABLE t1 (f1 INT PRIMARY KEY, f2 INT UNIQUE);
+	CREATE TABLE t1 (f1 INT PRIMARY KEY, f2 INT NOT NULL UNIQUE);
```
Comment (lysu) [inline executor/write_test.go]: maybe add some `generated col`/`auto inc col` with `null`/`!hasValue` case~?

File: executor/insert_common.go [hunk 9]
```diff
@@ -185,18 +187,19 @@ func (e *InsertValues) handleErr(col *table.Column, val *types.Datum, rowIdx int
...
+func (e *InsertValues) evalRow(cols []*table.Column, list []expression.Expression, rowIdx int) ([]types.Datum, error) {
 	rowLen := len(e.Table.Cols())
 	if e.hasExtraHandle {
 		rowLen++
 	}
 	row := make([]types.Datum, rowLen)
 	hasValue := make([]bool, rowLen)
 
+	// For statements like `insert into t set a = b + 1`.
 	if e.needFillDefaultValues {
```
Comment (lysu) [inline executor/insert_common.go]: maybe this fileld also need rename~?

File: executor/insert_common.go [hunk 10]
```diff
@@ -215,34 +218,36 @@ func (e *InsertValues) getRow(cols []*table.Column, list []expression.Expression
...
 //     2. for nullable and have default value column, use it's default value.
 //     3. for not null column, use zero value even in strict mode.
 //     4. for auto_increment column, use zero value.
 //     5. for generated column, use NULL.
-func (e *InsertValues) fillDefaultValues(row []types.Datum, hasValue []bool) error {
+func (e *InsertValues) setValueForRefColumn(row []types.Datum, hasValue []bool) error {
 	for i, c := range e.Table.Cols() {
-		var err error
-		if c.IsGenerated() {
-			continue
```
Comment (lysu) [inline executor/insert_common.go]: Is this behavier has be changed~? maybe need a test for  ref column generate insert ?

File: executor/insert_common.go (hunk 9 above)
Comment (jackysp) [reply executor/insert_common.go]: I don't think so.

File: executor/insert_common.go (hunk 10 above)
Comment (jackysp) [reply executor/insert_common.go]: I think we have a case for it. Let me check.

Comment (jackysp) [reply executor/insert_common.go]: ``getColDefaultValue` will return null for generated columns.

File: executor/insert_common.go (hunk 7 above)
Comment (jackysp) [reply executor/insert_common.go]: generated always doesn't have a value getColDefaultValue will return null for it.

File: executor/insert_common.go [hunk 11]
```diff
@@ -368,102 +352,116 @@ func (e *InsertValues) getColDefaultValue(idx int, col *table.Column) (d types.D
...
-		} else {
-			if !hasValue[i] || (mysql.HasNotNullFlag(c.Flag) && row[i].
-				IsNull() && e.ctx.GetSessionVars().StmtCtx.BadNullAsWarning) {
-				var err error
-				row[i], err = e.getColDefaultValue(i, c)
-				if e.filterErr(err) != nil {
-					return errors.Trace(err)
-				}
+			row[i], err = table.CastValue(e.ctx, val, c.ToInfo())
+			if err != nil {
```
[the remark this answers is not in the document]
Comment (jackysp) [reply executor/insert_common.go]: handleErr is not a good way to handle this. I'll fix it in another PR.

**Comment by jackysp** [comment]
PTAL @lysu

**Pull request merged**